services/QuillLMS/engines/evidence/app/services/evidence/gen_ai/repeated_feedback/checker.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module Evidence
+  module GenAI
+    module RepeatedFeedback
+      class Checker < ApplicationService
+        KEY_REPEAT = 'repeat_feedback'
+
+        attr_reader :feedback, :history, :optimal, :chat_api
+
+        def initialize(feedback:, history: [], optimal: false, chat_api: Evidence::OpenAI::Chat)
+          @feedback = feedback
+          @history = history
+          @optimal = optimal
+          @chat_api = chat_api
+        end
+
+        def run
+          return false if optimal
+          return false if history.empty? # can't be repeat with no history
+          return false if history.size >= 4 # No secondary feedback on last attempt
```

Comment:
Maybe extract a constant: `PENULTIMATE_ATTEMPT = 4` and get rid of comment?

Reply:
Good feedback. I've refactored this to make the code a little more obvious (and the comment something that the code doesn't give).